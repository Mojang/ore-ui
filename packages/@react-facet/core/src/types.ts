export type Immutable = boolean | number | string | undefined | null

export type ObjectWithImmutables = Record<string, Immutable>

export interface EqualityCheck<T> {
  (): (current: T) => boolean
}

/**
 * Possible values a facet can have.
 * Currently functions are not supported
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type Value = string | number | boolean | undefined | null | [] | Record<string, any> | bigint

export type ExtractFacetValues<T extends ReadonlyArray<Facet<unknown>>> = {
  [K in keyof T]: T[K] extends Facet<infer V> ? V : never
}

export const FACET_FACTORY = Symbol('facet-factory')

export interface FacetFactory<T> {
  (): Facet<T>
  factory: typeof FACET_FACTORY
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type ExcludeFacetFactory<T> = T extends FacetFactory<any> ? never : T

export interface Listener<T> {
  (value: T): void
}

export type Unsubscribe = () => void

export type Observe<T> = (listener: Listener<T>) => Unsubscribe

export interface Facet<T> {
  get: () => Option<T>
  observe: Observe<T>
}

export interface WritableFacet<T> extends Facet<T> {
  set: (value: T) => void
  setWithCallback: (callback: (previousValue: Option<T>) => Option<T>) => void
}

export interface Update<V> {
  (value: V): void
}

export type Setter<V> = (callbackSetter: V | ((previousValue: Option<V>) => Option<V>)) => void

export type Cleanup = Unsubscribe

export type StartSubscription<V> = (update: Update<V>) => Cleanup

export const isFacet = <T>(value: Value | Facet<T>): value is Facet<T> => {
  return (
    value !== null &&
    value !== undefined &&
    (value as Facet<T>).observe !== undefined &&
    (value as Facet<T>).get !== undefined
  )
}

export type FacetProp<T extends Value> = Facet<T> | ExcludeFacetFactory<T>

export interface Effect<T> {
  (value: T): void | Cleanup
}

export const NO_VALUE = Symbol.for('NoValue')

export type NoValue = typeof NO_VALUE

export type Option<T> = NoValue | T

/**
 * Unit of work that can be scheduled within a batch.
 */
export type Task = {
  (): void
  scheduled?: boolean
}

/**
 * Function that when executed will have its Facet updates scheduled.
 */
export type Batch = () => void

/**
 * CSSStyleDeclaration that has FacetDefinition as values and includes Gameface's extended number based properties
 * Based on Gameface's TypeScript definition: https://github.com/Mojang/Mojang_Cohtml/blob/master/src/Scripting/scripted/typescript/ScriptedCSSStyleDeclaration.d.ts
 */
export type FacetCSSStyleDeclaration = {
  alignContent?: FacetProp<string>
  alignItems?: FacetProp<string>
  alignSelf?: FacetProp<string>
  animation?: FacetProp<string>
  animationDelay?: FacetProp<string>
  animationDirection?: FacetProp<string>
  animationDuration?: FacetProp<string>
  animationFillMode?: FacetProp<string>
  animationIterationCount?: FacetProp<string>
  animationName?: FacetProp<string>
  animationPlayState?: FacetProp<string>
  animationTimingFunction?: FacetProp<string>
  background?: FacetProp<string>
  backgroundColor?: FacetProp<string>
  backgroundImage?: FacetProp<string>
  backgroundPosition?: FacetProp<string>
  backgroundRepeat?: FacetProp<string>
  backgroundSize?: FacetProp<string>
  border?: FacetProp<string>
  borderBottom?: FacetProp<string>
  borderBottomColor?: FacetProp<string>
  borderBottomLeftRadius?: FacetProp<string>
  borderBottomRightRadius?: FacetProp<string>
  borderBottomStyle?: FacetProp<string>
  borderBottomWidth?: FacetProp<string>
  borderColor?: FacetProp<string>
  borderImage?: FacetProp<string>
  borderImageOutset?: FacetProp<string>
  borderImageRepeat?: FacetProp<string>
  borderImageSlice?: FacetProp<string>
  borderImageSource?: FacetProp<string>
  borderImageWidth?: FacetProp<string>
  borderLeft?: FacetProp<string>
  borderLeftColor?: FacetProp<string>
  borderLeftStyle?: FacetProp<string>
  borderLeftWidth?: FacetProp<string>
  borderRadius?: FacetProp<string>
  borderRight?: FacetProp<string>
  borderRightColor?: FacetProp<string>
  borderRightStyle?: FacetProp<string>
  borderRightWidth?: FacetProp<string>
  borderStyle?: FacetProp<string>
  borderTop?: FacetProp<string>
  borderTopColor?: FacetProp<string>
  borderTopLeftRadius?: FacetProp<string>
  borderTopRightRadius?: FacetProp<string>
  borderTopStyle?: FacetProp<string>
  borderTopWidth?: FacetProp<string>
  borderWidth?: FacetProp<string>
  bottom?: FacetProp<string>
  bottomPX?: FacetProp<number>
  bottomVH?: FacetProp<number>
  bottomVW?: FacetProp<number>
  boxShadow?: FacetProp<string>
  clipPath?: FacetProp<string>
  color?: FacetProp<string>
  contain?: FacetProp<string>
  cursor?: FacetProp<string>
  display?: FacetProp<'flex' | 'none'>
  filter?: FacetProp<string>
  flex?: FacetProp<string>
  flexDirection?: FacetProp<string>
  flexWrap?: FacetProp<string>
  font?: FacetProp<string>
  fontFamily?: FacetProp<string>
  fontFitMode?: FacetProp<string>
  fontSize?: FacetProp<string>
  fontStyle?: FacetProp<string>
  fontWeight?: FacetProp<string>
  height?: FacetProp<string>
  heightPX?: FacetProp<number>
  heightVH?: FacetProp<number>
  isolation?: FacetProp<string>
  justifyContent?: FacetProp<string>
  left?: FacetProp<string>
  leftPX?: FacetProp<number>
  leftVH?: FacetProp<number>
  leftVW?: FacetProp<number>
  letterSpacing?: FacetProp<string>
  lineHeight?: FacetProp<string>
  margin?: FacetProp<string>
  marginBottom?: FacetProp<string>
  marginLeft?: FacetProp<string>
  marginRight?: FacetProp<string>
  marginTop?: FacetProp<string>
  mask?: FacetProp<string>
  maskClip?: FacetProp<string>
  maskImage?: FacetProp<string>
  maskMode?: FacetProp<string>
  maskPosition?: FacetProp<string>
  maskRepeat?: FacetProp<string>
  maskSize?: FacetProp<string>
  maxHeight?: FacetProp<string>
  maxHeightPX?: FacetProp<number>
  maxHeightVH?: FacetProp<number>
  maxWidth?: FacetProp<string>
  maxWidthPX?: FacetProp<number>
  maxWidthVW?: FacetProp<number>
  minHeight?: FacetProp<string>
  minHeightPX?: FacetProp<number>
  minHeightVH?: FacetProp<number>
  minWidth?: FacetProp<string>
  minWidthPX?: FacetProp<number>
  minWidthVW?: FacetProp<number>
  mixBlendMode?: FacetProp<string>
  opacity?: FacetProp<number>
  overflow?: FacetProp<string>
  overflowWrap?: FacetProp<string>
  overflowX?: FacetProp<string>
  overflowY?: FacetProp<string>
  padding?: FacetProp<string>
  paddingBottom?: FacetProp<string>
  paddingLeft?: FacetProp<string>
  paddingRight?: FacetProp<string>
  paddingTop?: FacetProp<string>
  perspective?: FacetProp<string>
  perspectiveOrigin?: FacetProp<string>
  pointerEvents?: FacetProp<string>
  position?: FacetProp<string>
  right?: FacetProp<string>
  rightPX?: FacetProp<number>
  rightVH?: FacetProp<number>
  rightVW?: FacetProp<number>
  textAlign?: FacetProp<string>
  textOverflow?: FacetProp<string>
  textShadow?: FacetProp<string>
  textTransform?: FacetProp<string>
  top?: FacetProp<string>
  topPX?: FacetProp<number>
  topVH?: FacetProp<number>
  topVW?: FacetProp<number>
  transform?: FacetProp<string>
  transformOrigin?: FacetProp<string>
  transition?: FacetProp<string>
  transitionDelay?: FacetProp<string>
  transitionDuration?: FacetProp<string>
  transitionProperty?: FacetProp<string>
  transitionTimingFunction?: FacetProp<string>
  userSelect?: FacetProp<string>
  visibility?: FacetProp<string>
  whiteSpace?: FacetProp<string>
  width?: FacetProp<string>
  widthPX?: FacetProp<number>
  widthVW?: FacetProp<number>
  zIndex?: FacetProp<number>
}
