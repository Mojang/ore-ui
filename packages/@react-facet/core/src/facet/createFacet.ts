import { defaultEqualityCheck } from '../equalityChecks'
import { Cleanup, EqualityCheck, Listener, WritableFacet, StartSubscription, Option, NO_VALUE } from '../types'
import { batch } from '../scheduler'

export interface FacetOptions<V> {
  initialValue: Option<V>
  startSubscription?: StartSubscription<V>
  equalityCheck?: EqualityCheck<V>
}
/**
 * The low level function to create a Facet, not recommended to be used if you can use any of the react facet hooks to create facets instead (Ex: useFacetState, useFacetWrap)
 */
export function createFacet<V>({
  initialValue,
  startSubscription,
  equalityCheck = defaultEqualityCheck,
}: FacetOptions<V>): WritableFacet<V> {
  const listeners: Set<Listener<V>> = new Set()
  let currentValue = initialValue
  let cleanupSubscription: Cleanup | undefined

  const checker = equalityCheck?.()

  const update = (newValue: V) => {
    if (equalityCheck !== undefined) {
      // we optimize for the most common scenario of using the defaultEqualityCheck (by inline its implementation)
      if (equalityCheck === defaultEqualityCheck) {
        const typeofValue = typeof newValue
        if (
          (typeofValue === 'number' ||
            typeofValue === 'string' ||
            typeofValue === 'boolean' ||
            newValue === null ||
            newValue === undefined) &&
          currentValue === newValue
        ) {
          return
        }
      } else {
        if (checker !== undefined && checker(newValue)) {
          return
        }
      }
    }

    batch(() => {
      currentValue = newValue

      for (const listener of listeners) {
        listener(currentValue)
      }
    })
  }

  /**
   * Simpler update implementation that only resets the value and runs all cleanup functions.
   * Done as a separated function to not interfere with the usual "hot-path" of the update function.
   */
  const updateToNoValue = () => {
    currentValue = NO_VALUE
  }

  return {
    set: update,

    setWithCallback: (setter) => {
      const value = setter(currentValue)
      if (value === NO_VALUE) {
        updateToNoValue()
      } else {
        update(value)
      }
    },

    get: () => currentValue,

    observe: (listener) => {
      listeners.add(listener)

      if (currentValue !== NO_VALUE) {
        listener(currentValue)
      }

      // This is the first subscription, so we start subscribing to dependencies
      if (listeners.size === 1 && startSubscription) {
        cleanupSubscription = startSubscription(update)
      }

      return () => {
        listeners.delete(listener)

        // if this was the last to unsubscribe, we unsubscribe from our dependencies
        if (listeners.size === 0 && cleanupSubscription) {
          currentValue = initialValue
          cleanupSubscription()
        }
      }
    },
  }
}
