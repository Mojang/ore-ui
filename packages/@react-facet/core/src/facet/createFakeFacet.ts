import { Facet } from '..'
/**
 * Creates a nonwritable barebones facet to be used when you need an initial facet value that's meant to be replaced later by a real facet. Ex: with `useContext()`
 */
export function createFakeFacet<T>(value: T): Facet<T> {
  const facet: Facet<T> = {
    get: () => value,
    observe: (listener) => {
      if (process.env.NODE_ENV !== 'production') {
        console.log(`Accessing a fake facet, perhaps you're missing a Context Provider?`)
      }
      listener(value)
      return () => {}
    },
  }
  return facet
}
