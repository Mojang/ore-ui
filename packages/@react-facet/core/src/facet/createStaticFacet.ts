import { Facet } from '../types'
/**
 * Creates a nonwritable barebones static facet to be used when you need an initial facet value outside the react context
 * that's meant to be replaced later by a real facet. Ex: with `createContext()`
 */
export function createStaticFacet<T>(value: T): Facet<T> {
  const facet: Facet<T> = {
    get: () => value,
    observe: (listener) => {
      listener(value)
      return () => {}
    },
  }
  return facet
}
