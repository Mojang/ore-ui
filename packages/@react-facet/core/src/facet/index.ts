export * from './createFacet'
export * from './createFacet'
