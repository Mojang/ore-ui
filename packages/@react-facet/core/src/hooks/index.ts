export * from './useFacetCallback'
export * from './useFacetEffect'
export * from './useFacetLayoutEffect'
export * from './useFacetMap'
export * from './useFacetMemo'
export * from './useFacetPropSetter'
export * from './useFacetRef'
export * from './useFacetRef'
export * from './useFacetState'
export * from './useFacetUnwrap'
export * from './useFacetWrap'
