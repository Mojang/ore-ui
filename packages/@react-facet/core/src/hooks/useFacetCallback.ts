import { useCallback, useRef } from 'react'
import { NoValue } from '..'
import { Facet, NO_VALUE, Option } from '../types'
import { useFacetEffect } from './useFacetEffect'

/**
 * Creates a callback that depends on the value of a facet.
 * Very similar to `useCallback` from `React`
 *
 * @param callback function callback that receives the current facet value and the arguments passed to the callback
 * @param dependencies variable used by the map that are available in scope (similar as dependencies of useCallback)
 * @param facet facet that the callback listens to
 *
 * We pass the dependencies of the callback as the second argument so we can leverage the eslint-plugin-react-hooks option for additionalHooks.
 * Having this as the second argument allows the linter to work.
 */
export function useFacetCallback<V, T>(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  callback: (value: V) => (...args: any[]) => T | NoValue,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  dependencies: any[],
  facet: Facet<V>,
) {
  const facetRef = useRef<Option<V>>(facet.get())

  useFacetEffect(
    (value) => {
      facetRef.current = value
    },
    [],
    facet,
  )
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const callbackMemoized = useCallback(callback, dependencies)

  return useCallback(
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (...args: any[]): T | NoValue => {
      const value = facetRef.current

      if (value === NO_VALUE) return NO_VALUE

      return callbackMemoized(value)(...args)
    },
    [callbackMemoized, facetRef],
  )
}
