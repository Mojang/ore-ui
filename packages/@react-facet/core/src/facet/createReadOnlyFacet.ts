import { Facet } from '../types'
import { createFacet, FacetOptions } from './createFacet'

export function createReadOnlyFacet<V>(options: FacetOptions<V>): Facet<V> {
  return createFacet(options)
}
