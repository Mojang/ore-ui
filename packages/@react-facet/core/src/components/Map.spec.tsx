import React from 'react'
import { act, render } from '@react-facet/dom-fiber-testing-library'
import { createFacet } from '../facet'
import { Facet } from '../types'
import { useFacetEffect, useFacetMap } from '../hooks'
import { Map } from '.'

it('renders all items in a Facet of array', () => {
  type Input = { a: string }
  const data = createFacet({
    initialValue: [{ a: '1' }, { a: '2' }, { a: '3' }, { a: '4' }, { a: '5' }],
  })

  const ExampleContent = ({ item }: { item: Facet<Input> }) => {
    return (
      <span>
        <fast-text text={useFacetMap(({ a }) => a, [], [item])} />
      </span>
    )
  }

  const inputEqualityCheck = () => {
    const previous: Partial<Input> = {}

    return (current: Input) => {
      if (current.a === previous.a) {
        return true
      }
      previous.a = current.a
      return false
    }
  }
  const Example = () => {
    return (
      <Map array={data} equalityCheck={inputEqualityCheck}>
        {(item) => <ExampleContent item={item} />}
      </Map>
    )
  }

  const scenario = <Example />

  const { container } = render(scenario)

  expect(container).toMatchSnapshot()
})

it('unmounts components when the array reduces in size', () => {
  interface Item {
    value: string
  }

  const data = createFacet<Item[]>({
    initialValue: [{ value: '1' }, { value: '2' }, { value: '3' }, { value: '4' }, { value: '5' }],
  })

  const Item = ({ item }: { item: Facet<Item> }) => (
    <span>
      <fast-text text={useFacetMap(({ value }) => value, [], [item])} />
    </span>
  )

  const itemEqualityCheck = () => {
    const previous: Partial<Item> = {}

    return (current: Item) => {
      if (current.value === previous.value) {
        return true
      }

      previous.value = current.value

      return false
    }
  }

  const Example = () => {
    return (
      <Map array={data} equalityCheck={itemEqualityCheck}>
        {(item) => <Item item={item} />}
      </Map>
    )
  }

  const scenario = <Example />

  const { container } = render(scenario)

  expect(container).toMatchSnapshot()

  data.set([{ value: '1' }, { value: '2' }])

  expect(container).toMatchSnapshot()
})

it('updates only items that have changed', () => {
  type Input = { a: string }
  const data = createFacet({
    initialValue: [{ a: '1' }, { a: '2' }, { a: '3' }, { a: '4' }, { a: '5' }],
  })

  const mock = jest.fn()

  const ExampleContent = ({ item }: { item: Facet<Input> }) => {
    useFacetEffect(mock, [], [item])
    return null
  }

  const inputEqualityCheck = () => {
    const previous: Partial<Input> = {}

    return (current: Input) => {
      if (current.a === previous.a) {
        return true
      }

      previous.a = current.a

      return false
    }
  }

  const Example = () => {
    return (
      <Map array={data} equalityCheck={inputEqualityCheck}>
        {(item) => <ExampleContent item={item} />}
      </Map>
    )
  }

  const scenario = <Example />

  render(scenario)

  expect(mock).toHaveBeenCalledTimes(5)

  mock.mockClear()

  act(() => {
    data.set([{ a: '6' }, { a: '2' }, { a: '3' }, { a: '4' }, { a: '5' }])
  })

  expect(mock).toHaveBeenCalledTimes(1)
  expect(mock).toHaveBeenCalledWith({ a: '6' })
})
