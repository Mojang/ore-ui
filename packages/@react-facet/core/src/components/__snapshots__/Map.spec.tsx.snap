// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders all items in a Facet of array 1`] = `
<div>
  <span>
    1
  </span>
  <span>
    2
  </span>
  <span>
    3
  </span>
  <span>
    4
  </span>
  <span>
    5
  </span>
</div>
`;

exports[`unmounts components when the array reduces in size 1`] = `
<div>
  <span>
    1
  </span>
  <span>
    2
  </span>
  <span>
    3
  </span>
  <span>
    4
  </span>
  <span>
    5
  </span>
</div>
`;

exports[`unmounts components when the array reduces in size 2`] = `
<div>
  <span>
    1
  </span>
  <span>
    2
  </span>
</div>
`;
